js/components/controlButton.js

```diff
@@ -0,0 +1,316 @@
+    var responsiveUiMultiple = (breakpoints[responsiveView] || {}).multiplier || 1;
+    return responsiveUiMultiple;
+  },
+
+  /**
+   * Determines whether or not tooltips are enabled considering the current
+   * platform and skin configuration.
+   * @private
+   * @return {Boolean} True if tooltips are enabled, false otherwise
+   */
+  areTooltipsEnabled: function() {
+    var enabled = false;
+
+    if (!this.props.controller.state.isMobile) {
+      enabled = Utils.getPropertyValue(
+        this.props.skinConfig,
+        'controlBar.tooltips.enabled',
+        false
+      );
+    }
+    return enabled;
+  },
+
+  /**
+   * Determines the vertical offset value to use for tooltips depending on the
+   * props that were passed to this component.
+   * @private
+   * @return {Number} A numerical value representing the vertical offset at which tooltips will be rendered
+   */
+  getTooltipVerticalOffset: function() {
+    var tooltipVerticalOffset;
+    // Use tooltipVerticalOffset if provided, otherwise use control bar height as default
+    if (typeof this.props.tooltipVerticalOffset !== 'undefined') {
+      tooltipVerticalOffset = this.props.tooltipVerticalOffset;
+    } else {
+      tooltipVerticalOffset = Utils.getPropertyValue(
+        this.props.skinConfig,
+        'controlBar.height',
+        0
+      );
+    }
+    return tooltipVerticalOffset;
+  },
+
+  /**
+   * Either executes a callback passed by the parent that determines the tooltip
+   * alignment or returns the default tooltip alignment value.
+   * @private
+   * @param {type} key An id (usually the focusId prop) that identifies the button whose tooltip alignment we want to determine
+   * @return {string} A value from CONSTANTS.TOOLTIP_ALIGNMENT which represents the tooltip alignment
+   */
+  getTooltipAlignment: function(key) {
+    if (typeof this.props.getTooltipAlignment === 'function') {
+      return this.props.getTooltipAlignment(key);
+    } else {
+      return CONSTANTS.TOOLTIP_ALIGNMENT.CENTER;
+    }
+  },
+
+  /**
+   * Handler for the mouseenter event.
+   * @private
+   * @param {Event} The mouseenter event object
+   */
+  onMouseEnter: function(event) {
+    this.highlight(event);
+
+    if (typeof this.props.onMouseOver === 'function') {
+      this.props.onMouseOver(event);
+    }
+  },
+
+  /**
+   * Handler for the mouseleave event.
+   * @private
+   * @param {Event} The mouseleave event object
+   */
+  onMouseLeave: function(event) {
+    this.removeHighlight(event);
+
+    if (typeof this.props.onMouseOut === 'function') {
+      this.props.onMouseOut(event);
+    }
+  },
+
+  render: function() {
+    var Component = this.getButtonComponent();
+    var className = classNames('oo-control-bar-item', this.props.className);
+    var iconStyles = this.getIconStyles();
+    var areTooltipsEnabled = this.areTooltipsEnabled();
+
+    if (areTooltipsEnabled && this.props.tooltip) {
+      var responsiveUiMultiple = this.getResponsiveUiMultiple();
```

Comment:
Should these be declared outside of this if statement with a default value?

Reply:
I can update this if you prefer it, but I think it's a bit cleaner this way. These don't need a default value when tooltips aren't rendered, so the automatic hoisting achieves the same result. I'm ok with changing this though.

Reply:
Either way is fine. The default value was the main point rather than the hoisting